sopel/coretasks.py

```diff
@@ -893,175 +1001,144 @@ def track_quit(bot, trigger):
+        LOGGER.info('No capability negotiation.')
+        bot.write(('CAP', 'END'))
 
-    # Not the last in a multi-line reply. First two args are * and LS.
-    if trigger.args[2] == '*':
-        return
 
-    LOGGER.info(
-        "Client capability negotiation list: %s",
-        ', '.join(batched_caps.keys()),
-    )
-    bot.server_capabilities = batched_caps
-
-    # If some other plugin requests it, we don't need to add another request.
-    # If some other plugin prohibits it, we shouldn't request it.
-    core_caps = [
-        'echo-message',
-        'multi-prefix',
-        'away-notify',
-        'chghost',
-        'cap-notify',
-        'server-time',
-        'userhost-in-names',
-        'message-tags',
-    ]
-    for cap in core_caps:
-        if cap not in bot._cap_reqs:
-            bot._cap_reqs[cap] = [utils.CapReq('', 'coretasks')]
-
-    def acct_warn(bot, cap):
-        LOGGER.info("Server does not support %s, or it conflicts with a custom "
-                    "plugin. User account validation unavailable or limited.",
-                    cap[1:])
-        if bot.config.core.owner_account or bot.config.core.admin_accounts:
-            LOGGER.warning(
-                "Owner or admin accounts are configured, but %s is not "
-                "supported by the server. This may cause unexpected behavior.",
-                cap[1:])
-    auth_caps = ['account-notify', 'extended-join', 'account-tag']
-    for cap in auth_caps:
-        if cap not in bot._cap_reqs:
-            bot._cap_reqs[cap] = [utils.CapReq('', 'coretasks', acct_warn)]
-
-    for cap, reqs in bot._cap_reqs.items():
-        # At this point, we know mandatory and prohibited don't co-exist, but
-        # we need to call back for optionals if they're also prohibited
-        prefix = ''
-        for entry in reqs:
-            if prefix == '-' and entry.prefix != '-':
-                entry.failure(bot, entry.prefix + cap)
-                continue
-            if entry.prefix:
-                prefix = entry.prefix
-
-        # It's not required, or it's supported, so we can request it
-        if prefix != '=' or cap in bot.server_capabilities:
-            # REQs fail as a whole, so we send them one capability at a time
-            bot.write(('CAP', 'REQ', entry.prefix + cap))
-        # If it's required but not in server caps, we need to call all the
-        # callbacks
-        else:
-            for entry in reqs:
-                if entry.failure and entry.prefix == '=':
-                    entry.failure(bot, entry.prefix + cap)
+def _handle_cap_acknowledgement(
+    bot: SopelWrapper,
+    cap_req: Tuple[str, ...],
+    results: List[Tuple[bool, Optional[plugin.CapabilityNegotiation]]],
+    was_completed: bool,
+) -> None:
+    if any(
+        callback_result[1] == plugin.CapabilityNegotiation.ERROR
+        for callback_result in results
+    ):
+        # error: a plugin needs something and the bot cannot function properly
```

Comment:
The problem here is that now, the plugin *is* loaded, and its callables are active. If a plugin cannot properly handle CAP REQ, and either crash or explicitly returns `CapabilityNegotiation.ERROR`, I think it is best for the bot to stop here and let the bot owner figure out what is what.

Beside, this should be a very rare scenario. Remember that most plugins just don't use CAP REQ. We are working with edge cases here.

Reply:
It is doable, yes, and yet I'm reluctant to allow a bot to run with a broken plugin that soon in the launch process: unlike many other features, capability negotiation can have a big impact on the server & client's behaviors. As a result, I think there are bigger risks here for security.

For instance, if something mess up with the SASL authentication, or any kind of authentication, a bot owner may not realize that its bot instance is not secure at all, because the plugin that was supposed to handle that with CAP REQ is not loaded at all. It's unlikely to be a problem, but the risk is here.

In that case, I don't think Sopel should unload the plugin and carry on, I think it should tell the bot owner that something is wrong with their setup, and they must act on it.